File: tcpcom/EV3/Java/Examples/TrainServer.java
// TrainServer.java

import ch.aplu.ev3.*;
import ch.aplu.tcpcom.*;

public class TrainServer implements TCPServerListener
{
  private TCPServer server;
  private LegoRobot robot;
  private Gear gear;
  private boolean isWaiting = true;
  private long startTime;

  public TrainServer()
  {
    robot = new LegoRobot();
    gear = new Gear();
    robot.addPart(gear);
    gear.setSpeed(25);
    LightSensor lsBottom = new LightSensor(SensorPort.S3);
    robot.addPart(lsBottom);
    LightSensor lsFront = new LightSensor(SensorPort.S1);
    robot.addPart(lsFront);
    int port = 5000;
    server = new TCPServer(port);
    server.addTCPServerListener(this);
    robot.drawString("Train Server", 0, 1);
    startTime = System.nanoTime();
    while (!robot.isEscapeHit())
    {
      if (isWaiting)
         gear.stop();
      int vB = lsBottom.getValue();
      int vF = lsFront.getValue();
      if (vB < 500)
        gear.leftArc(0.10);
      else
        gear.rightArc(0.10);
      if (!isWaiting && vF > 500 && System.nanoTime() - startTime > 5000000000L)
      {
        server.sendMessage("go");
        robot.drawString("Stopped", 0, 2);
        isWaiting = true;
      }
    }
    server.terminate();
    robot.exit();
  }

  public void onStateChanged(String state, String msg)
  {
    if (state.equals(TCPServer.LISTENING))
    {
      robot.drawString("Stopped", 0, 2);
      isWaiting = true;
    }
    else if (state.equals(TCPServer.MESSAGE))
    {
      isWaiting = false;
      startTime = System.nanoTime();
      robot.drawString("Running", 0, 2);
    }
  }

  public static void main(String[] args)
  {
    new TrainServer();
  }
}

File: tcpcom/Pi2Go/TigerJython/Examples/TrainServer.py
# TrainServer.py

from raspibrick import *
from tcpcom import TCPServer

def onStateChanged(state, msg):
    global isWaiting, start
    if state == TCPServer.LISTENING:
        ledLeft.setColor("red")
        display.showText("stop")
        gear.stop()
        isWaiting = True
    if state == TCPServer.CONNECTED:
        ledLeft.setColor("green")
    if state == TCPServer.MESSAGE:
        isWaiting = False
        start = time.time()
        display.showText("run")

robot = Robot()
display = Display()
ledLeft = Led(LED_LEFT)
ledLeft.setColor("red")
gear = Gear()
gear.setSpeed(25)
irLeft = InfraredSensor(IR_LINE_LEFT)
irRight = InfraredSensor(IR_LINE_RIGHT)
irCenter = InfraredSensor(IR_CENTER)
r = 0.1
port = 5000
server = TCPServer(port, stateChanged = onStateChanged)
start = time.time()
isWaiting = True
while not isEscapeHit():
    if isWaiting:
        continue
    vL = irLeft.getValue()
    vR = irRight.getValue()
    vC = irCenter.getValue()
    if vL == 1 and vR == 0:
        gear.forward()
    elif vL == 0 and vR == 0:   
        gear.leftArc(r)
    elif vL == 1 and vR == 1:
        gear.rightArc(r)
    if vC == 1 and time.time() - start > 5:
        server.sendMessage("go")
        display.showText("stop")
        gear.stop()
        isWaiting = True
                
server.terminate()        
robot.exit()
